templates/ingress.yaml

```diff
@@ -73,29 +79,46 @@ spec:
   {{- end }}
   rules:
   {{- range $ingressRuleIndex, $ingressRule := $.Values.ingress.rules }}
+
+  {{- $svcPort := 0 }}
+  {{- if $ingressRule.servicePort }}
+  {{- $svcPort = $ingressRule.servicePort -}}
+  {{- else }}
+  {{- $svcPort = (index $.Values.service.ports 0).port -}}
+  {{- end }}
+
+  {{- $svcName := "" }}
+  {{- if $ingressRule.serviceName }}
+  {{- $svcName = $ingressRule.serviceName -}}
+  {{- else }}
+  {{- if $.Values.service.name }}
+  {{- $svcName = $.Values.service.name -}}
+  {{- else }}
+  {{- $svcName = $.Release.Name -}}
+  {{- end }}
+  {{- end }}
   - host: {{ $ingressRule.host | quote }}
     http:
       paths:
       - backend:
-          {{- if $ingressRule.serviceName }}
-          serviceName: {{ $ingressRule.serviceName | quote }}
-          {{- else }}
-          {{- if $.Values.service.name }}
-          serviceName: {{ $.Values.service.name | quote }}
-          {{- else }}
-          serviceName: {{ $.Release.Name | quote }}
-          {{- end }}
-          {{- end }}
-          {{- if $ingressRule.servicePort }}
-          servicePort: {{ $ingressRule.servicePort }}
-          {{- else }}
-          servicePort: {{ (index $.Values.service.ports 0).port }}
-          {{- end }}
+        {{- if semverCompare ">=1.19-0" $.Capabilities.KubeVersion.GitVersion }}
+          service:
+            name: {{ $svcName | quote }}
+            port:
+              number: {{ $svcPort }}
+        {{- else}}
+          serviceName: {{ $svcName }}
+          servicePort: {{ $svcPort }}
+        {{- end }}
         {{- if $ingressRule.path }}
         path: {{ $ingressRule.path | quote }}
         {{- else }}
         path: /
         {{- end }}
+        {{- if and $ingressRule.pathType (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion) }}
+        pathType: {{ $ingressRule.pathType | quote }}
+        {{- end }}
+        
```

Comment:
Same as above. Preferably using `Capabilities.APIVersions.Has`

Reply:
It's important to check kube version here, not api version
Because `pathType` was introduced only in v1.18 in v1beta1/Ingress

kube v1.17 v1beta1/Ingress - no pathType:
https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#httpingresspath-v1beta1-extensions
kube v1.18 v1beta1/Ingress - pathType required:
https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#httpingresspath-v1beta1-extensions